### ISSUE TITLE
ECS Agent disconnected problem

### ISSUE THREAD

**Issue by rossf7** [aws/amazon-ecs-agent#103]
We're using ECS for [force12.io](http://force12.io) our demo of micro scaling. We're seeing intermittent problems when one of our container instances stops responding for between 30 and 60 seconds. During this time the agent connected flag in the ECS web console is false. We also see the 2 errors below in the agent logs.

```
2015-06-08T15:06:09Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:06:09Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.20.16:443: use of closed network connection"
```

This error is occuring 6 or 7 times an hour on each container instance. Here is when it occurred between 15:00 and 16:00 UTC today.

```
2015-06-08T15:06:09Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:06:09Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:14:49Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.18.202:443: use of closed network connection"
2015-06-08T15:14:49Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.18.202:443: use of closed network connection"
2015-06-08T15:23:31Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:23:31Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:32:08Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:32:08Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:39:20Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:39:20Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:48:08Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:48:08Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.19.128:443: use of closed network connection"
2015-06-08T15:55:56Z [ERROR] Error getting message from acs module="acs client" err="read tcp 54.239.20.16:443: use of closed network connection"
2015-06-08T15:55:56Z [INFO] Error from acs; backing off module="acs handler" err="read tcp 54.239.20.16:443: use of closed network connection"
```

I've uploaded the full [agent log](https://s3.amazonaws.com/force12-ecs-logs/ecs-agent.log.2015-06-08-15.gz) to S3 for this hour. 
Our instances are running amzn-ami-2015.03.b-amazon-ecs-optimized - ami-d0b9acb8

Please let me know if you need any further information or additional logs.

Thanks

Ross

**Comment by euank** [comment]
Thanks for reporting the issue, Ross.
It manifests for me fairly reliably when I start tasks quickly (at least 1 a minute) for extended periods.
I think I figured out what's happening and I'm working on a fix (see above).

Again, thanks for reporting!
Best,
Euan

**Comment by rossf7** [comment]
Hi Euan,
great, thanks for this. We're happy to test the fix in the dev branch once its ready if that is helpful.
Cheers

Ross

**Comment by euank** [comment]
@rossf7 apologies for the delay in getting that merged. It's now in the current dev branch and should be in the next release. If you'd like to test it, that'd be awesome!

Thanks again for reporting the issue and bearing with me,
Euan

**Comment by rossf7** [comment]
Hi @euank,
great, thanks for this! I've done some testing and it looks much better.

I've built the agent from the dev branch and deployed it to our [staging rig](http://staging.force12.io). The 3 container instances are running CoreOS-beta-717.1.0-hvm (ami-7f9a6214).

I'm still seeing the warning message below but I think this is expected?

```
2015-07-13T17:06:53Z [WARN] Could not submit a container state change module="api client" change="{TaskArn:arn:aws:ecs:us-east-1:187012023547:task/634649e8-47c1-4de3-b56e-915afd911166 ContainerName:priority1 Status:4 Reason: ExitCode:0xc2089fe068 PortBindings:[] SentStatus:RUNNING}" err="Post https://ecs.us-east-1.amazonaws.com/: read tcp 54.239.20.154:443: use of closed network connection"
```

We're no longer seeing the problem where the agents stop responding. The demo itself is now tracking our random demand metric most of the time. I'll continue to monitor over the next couple of days.

Thanks

Ross

**Comment by samuelkarp** [comment]
We've released [v1.3.1](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.3.1), which we believe has addressed this problem.  Please feel free to reopen or file a new issue if the problem persists.

Ross, that warning should be benign as the Agent should open a new connection and retry the submission.

Thanks!
Sam
